File: LowestCommonAncestor.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */

public class Solution {
    
    public List<TreeNode> tree, branch;
    
    TreeNode LCA = null;
    
    public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) 
    {
        tree = new List<TreeNode>();
        recordTreePreOrder(root, tree);
        
        for (int i = 0; i < tree.Count; i++)
        {

            branch = new List<TreeNode>();
            recordTreePreOrder(tree[i], branch);
            
            if ((branch.Contains(p)) && (branch.Contains(q)))
            {
                LCA = tree[i];
            }
        }
    
        return LCA; 
    }
    
    public void recordTreePreOrder(TreeNode node, List<TreeNode> path)
    {        
        if (node != null)
        {
            path.Add(node);
            recordTreePreOrder(node.left, path);
            recordTreePreOrder(node.right, path);   
        }
    }
}

File: Python/lowestCommonAncestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':

        lca = None

        def digDown(root):

            nonlocal lca

            if not(root):
                return

            digDown(root.left)
            digDown(root.right)

            if p.val > q.val:
                if ((p.val >= root.val) and (q.val <= root.val)):
                    lca = root
            
            else:
                if ((p.val <= root.val) and (q.val >= root.val)):
                    lca = root
        
        digDown(root)
        return lca
